### ISSUE TITLE
[QST] Are cudf and nvtabular limited by total Physical CPU RAM + GPU RAM? Workaround for large dataset provisioning?

### ISSUE THREAD

**Issue by lightsailpro** [NVIDIA-Merlin/Transformers4Rec#432]
# ❓ Questions & Help

## Details

I have raw data in csv that is big (37GB). I have 64GB CPU ram + plus a V100 GPU with 16GB ram. The cudf.read_csv threw out memory GPU XID error even with cudf.set_allocator("managed") settings. While nvtableau and cudf can speed up dataframe operation using GPU(s), but they can not shuffle to files (as other frameworks e.g. hive, spark) during processing and are limited by physical GPU ram, is this right assessment? Any suggestion for large dataset that exceeds GPU RAM? Maybe use hive/spark to generate train/test/validation parquet files?

Also the training step is not limited by the GPU ram since batch size can be adjusted, is this correct?

**Comment by rnyak** [comment]
@lightsailpro cuDF works on single GPU and consumes GPU memory. So it is expected to get OOM error from a 37GB dataset on a 16GB GPU machine. We recommend you to use use `NVTabular`  for your processing your 37 GB dataset to avoid OOM issues on a single 16GB GPU. Note that NVTabular is using `dask-cudf` under the hood, which is doing distributed and parallel computing.  If you share your NVTabular code snippet we can help you better.  Please check out our [examples](https://github.com/NVIDIA-Merlin/NVTabular/tree/main/examples) to learn more about NVTabular pipeline.

For shuffling questions, yes,  NVTabular has `shuffle` feature and can shuffle the processed files before writing them to disk. See this unit test as an [example](https://github.com/NVIDIA-Merlin/NVTabular/blob/main/tests/unit/workflow/test_workflow.py#L152-L156). Note that NVTabular dataloaders has `shuffle` argument as well.

What model you would like to train? an MLP model? or something else? Which framework you use? TF or PyT or something else?  For model training please checkout Merlin Models [examples](https://github.com/NVIDIA-Merlin/models/tree/main/examples). 

Hope that helps.

**Comment by EvenOldridge** [comment]
If you convert your code to parquet then the parquet reader can read in and iterate over partitions which allows the GPU to pull in a subset of the data rather than the whole thing.  NVTabular will do that conversion for you.

**Comment by lightsailpro** [comment]
Thanks @rnyak and @EvenOldridge for the suggestions. I am basically following the template in this Tutorial: End-to-end Session-based Recommendation (https://github.com/NVIDIA-Merlin/Transformers4Rec/blob/main/examples/tutorial/02-ETL-with-NVTabular.ipynb) for a traditional sequential next item scenario with one user one session. My dataset is much larger though. I was able to workaround the cudf RAM issue by reducing number of category columns in my raw data and using pandas to read the csv, then without further maneuver and write to parquet file directly from pandas. Then use the following without explicitly using cudf data frame, and save the workflow processed to multiple parquet files. Could not use the save_time_based_splits to generate train/test/valid files either since it threw out OOM error too. I plan to use one of the eight parquet files as validation input, the rest files for training. I do not use day index. Not sure if there is any flaw here with my workaround. Please let me know if it is wrong.

Ideally if the code in step1 and 2 preprocessing/tabular etl including save_time_based_splits can be ported to hive SQL / or spark scalar code and generate the train/valid/test files from a raw csv files, I feel it will be more scalable on the data provision side for shops with only one or two low end GPUs.
...
dataset = nvt.Dataset(os.path.join(INPUT_DATA_DIR, 'test1.parquet'), engine='parquet', part_size='500MB', part_mem_fraction=0.15)
workflow = nvt.Workflow(filtered_sessions)
shuffle = Shuffle.PER_WORKER  # Shuffle algorithm
out_files_per_proc = 8  # Number of output files per worker
workflow.fit_transform(dataset).to_parquet(
    output_path="/test1_processed",
    shuffle=shuffle,
    out_files_per_proc=out_files_per_proc,
)

**Comment by lightsailpro** [comment]
@rnyak , Thanks. Yes, I am basically using the same same NVT pipeline given in [02-ETL-with-NVTabular.ipynb] without day index.
I could not use save_time_based_splits() since it threw out OOM error. I believe it uses cudf inside the code. In my case with the large dataset and a single 16GB GPU, It seems that anything using cudf will cause OOM error. "workflow.fit_transform(dataset).to_parquet" was able to generate 8 parquet files. I loop through these 8 files for training and validation. Unfortunately I had to set data_loader_engine='pyarrow'. With nvtabular as data loader, the training step failed around half way through. I only get about 4 iterations / sec with batch size 16 and max_sequence_length = 300, but I expect it being able to finish the training step eventually.

**Comment by rnyak** [comment]
@lightsailpro the easiest thing to do is to just edit the `save_time_based_splits()` function. you can either use pandas, or dask_cudf.

**Comment by alan-ai-learner** [comment]
save_time_based_splits(), is using my complete ram, i'm trying to use a parquet file of size 800MB. I have a RAM of 25gb.
Any workaround will be helpful.
